app/geant-exporter/geant-exporter.cc

```diff
@@ -0,0 +1,246 @@
+    /*!
+     * Storing both particle definition information and physics tables
+     *
+     * Code was splitted into two separate loops for simplicity:
+     * - First loop stores particle info; second stores physics tables
+     * - Minimal runtime difference penalty in favor of an easier readability
+     */
+    //-----------------------------------------------------------------------//
+
+    auto particle_iterator = G4ParticleTable::GetParticleTable()->GetIterator();
+
+    //-----------------------------------------------------------------------//
+    /*!
+     * LOOP 1: Storing particle information
+     */
+    //-----------------------------------------------------------------------//
+
+    // Creating the particle definition root file
+    G4String               root_output_filename = "geant-exporter-data.root";
+    std::unique_ptr<TFile> root_output(
+        new TFile(root_output_filename.c_str(), "recreate"));
+
+    TDirectory* particles = root_output->mkdir("particles");
+    particles->cd();
+
+    std::cout << "Creating " << root_output_filename << "..." << std::endl;
+
+    // Creating variables to reference the root branches
+    std::string name;
+    int         pdg;
+    double      mass, charge, lifetime, spin;
+    bool        is_stable;
+
+    particle_iterator->reset();
+
+    while ((*particle_iterator)())
+    {
+        G4ParticleDefinition* particle = particle_iterator->value();
+
+        //-------------------------------------------------------------------//
+        /*!
+         * Skipping the Geantino
+         *
+         * Friendly reminder that GenericIon and geantino both have the same
+         * pdg encoding (0). This will crash Celeritas during the creation
+         * of its ParticleParams, which will treat them as a duplicate entry.
+         */
+        //-------------------------------------------------------------------//
+        if (particle->GetParticleName() == "geantino")
+            continue;
+
+        name      = particle->GetParticleName();
+        pdg       = particle->GetPDGEncoding();
+        mass      = particle->GetPDGMass();
+        charge    = particle->GetPDGCharge();
+        lifetime  = particle->GetPDGLifeTime();
+        spin      = particle->GetPDGSpin();
+        is_stable = particle->GetPDGStable();
+
+        // Converting lifetime of unstable particles to seconds
+        // For stable particles lifetime = -1
+        if (!is_stable)
+        {
+            lifetime *= 1E-9;
+        }
+
+        // ROOT does not like special characters for its TTree names
+        replace_characters(name, "+", "Plus");
+        replace_characters(name, "-", "Minus");
+
+        // Gamma is a ROOT function. A TTree with that same name confuses ROOT
+        if (name == "gamma")
+        {
+            name = "photon";
+        }
+
+        // Creating a TTree for each particle
+        TTree* tree_particle = new TTree(name.c_str(), name.c_str());
```

Comment:
In ROOT, there is a sort of implicit shared ownership of some objects (TTree and TH1*).  The new interface will replace this implicit shared ownership with an explicit ownership model (using unique_ptr and shared_ptr as appropriate).

For the existing interfaces (e.g. this TTree object) you _can_ use a unique_ptr and things will be fine (```~TTree``` broadcast to other shared owner that the deletion happened).   You can even create the TTree on the stack if needed.   

Two things to watch out: 

one is then to not delete the shared owner (usually the TFile) before the unique_ptr goes out of scope (because the TFile will delete the TTree but (unless you register to be thus inform) the local unique_ptr is not reset and thus point to deleted memory).

Second, if you are deleting the TTree object yourself, then you also need to call the 'Write' for it explicitly.